### PR TITLE
[MNG-7830] Switch from plexus-xml to stax / woodstox

### PR DESCRIPTION
**Pull request by gnodet**
JIRA issue: https://issues.apache.org/jira/browse/MNG-7830
IT PR: https://github.com/apache/maven-integration-testing/pull/274

Switch the underlying plexus-xml Xpp3 parser to the STAX api / woodstox parser.
Woodstox Stax parser is 20% faster than the xpp3 parser.

The PR also moves out of modello generated code for the repository metadata and core extensions xml models.

Some pom are not valid xml files, for example:
```
Unable to parse using stax: /Users/gnodet/.m2/repository/org/eclipse/org.eclipse.osgi/3.8.0.v20120529-1548/org.eclipse.osgi-3.8.0.v20120529-1548.pom (Undeclared namespace prefix "xsi" (for attribute "schemaLocation")
 at [row,col,system-id]: [1,108,"/Users/gnodet/.m2/repository/org/eclipse/org.eclipse.osgi/3.8.0.v20120529-1548/org.eclipse.osgi-3.8.0.v20120529-1548.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/org/eclipse/equinox/org.eclipse.equinox.coordinator/1.1.0.v20120522-1841/org.eclipse.equinox.coordinator-1.1.0.v20120522-1841.pom (Undeclared namespace prefix "xsi" (for attribute "schemaLocation")
 at [row,col,system-id]: [1,108,"/Users/gnodet/.m2/repository/org/eclipse/equinox/org.eclipse.equinox.coordinator/1.1.0.v20120522-1841/org.eclipse.equinox.coordinator-1.1.0.v20120522-1841.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/org/eclipse/equinox/org.eclipse.equinox.region/1.2.101.v20150831-1342/org.eclipse.equinox.region-1.2.101.v20150831-1342.pom (Undeclared namespace prefix "xsi" (for attribute "schemaLocation")
 at [row,col,system-id]: [1,108,"/Users/gnodet/.m2/repository/org/eclipse/equinox/org.eclipse.equinox.region/1.2.101.v20150831-1342/org.eclipse.equinox.region-1.2.101.v20150831-1342.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/org/apache/hadoop/hadoop-project/3.2.2/hadoop-project-3.2.2.pom (Unexpected character '-' (code 45) (expected a name start character)
 at [row,col,system-id]: [1942,12,"/Users/gnodet/.m2/repository/org/apache/hadoop/hadoop-project/3.2.2/hadoop-project-3.2.2.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/activation/jakarta.activation/2.0.0/jakarta.activation-2.0.0.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [76,16,"/Users/gnodet/.m2/repository/com/sun/activation/jakarta.activation/2.0.0/jakarta.activation-2.0.0.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/activation/jakarta.activation/2.0.0-RC3/jakarta.activation-2.0.0-RC3.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [76,16,"/Users/gnodet/.m2/repository/com/sun/activation/jakarta.activation/2.0.0-RC3/jakarta.activation-2.0.0-RC3.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M5/jaxb-parent-3.0.0-M5.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [436,45,"/Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M5/jaxb-parent-3.0.0-M5.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M3/jaxb-parent-3.0.0-M3.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [434,45,"/Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M3/jaxb-parent-3.0.0-M3.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M4/jaxb-parent-3.0.0-M4.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [434,45,"/Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0-M4/jaxb-parent-3.0.0-M4.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0/jaxb-parent-3.0.0.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [436,45,"/Users/gnodet/.m2/repository/com/sun/xml/bind/mvn/jaxb-parent/3.0.0/jaxb-parent-3.0.0.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/istack/istack-commons/4.0.0-M2/istack-commons-4.0.0-M2.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [418,37,"/Users/gnodet/.m2/repository/com/sun/istack/istack-commons/4.0.0-M2/istack-commons-4.0.0-M2.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/istack/istack-commons/4.0.0-M3/istack-commons-4.0.0-M3.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [426,45,"/Users/gnodet/.m2/repository/com/sun/istack/istack-commons/4.0.0-M3/istack-commons-4.0.0-M3.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/istack/istack-commons/3.0.10/istack-commons-3.0.10.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [408,37,"/Users/gnodet/.m2/repository/com/sun/istack/istack-commons/3.0.10/istack-commons-3.0.10.pom"])
Unable to parse using stax: /Users/gnodet/.m2/repository/com/sun/istack/istack-commons/3.0.11/istack-commons-3.0.11.pom (Undeclared namespace prefix "Xlint"
 at [row,col,system-id]: [408,37,"/Users/gnodet/.m2/repository/com/sun/istack/istack-commons/3.0.11/istack-commons-3.0.11.pom"])
```
The `StaxTest` unit test (disabled by default) reads all POMs in the local repository.
We could disable namespace support when reading the POMs, but I'm not sure that's a good idea.

### CODE REVIEW

**Comment by gnodet** [comment]
> Wow. This looks great. definitely a big leap forward.
> 
> I might have missed it, but I didn't find any explicit use of the Stax2 API or Woodstox classes. Could this be done with JDK classes only? classes?

Stax2 / woodstox is actually needed for the consumer POM transformation: the stax api is not sufficient and does not allow full round tripping with xml as spaces in prolog are not reported.  Even aalto-xml does not support it.  The effect is that when using another implementation, the line breaks before the first element of the POM are removed, so the generated POM will usually contains the xml declaration, the license and the `<project>` element on a single line.  Most importantly, this breaks tests :-)

Apart from this use case, changing the implementation leads to various small issues as they do sometimes slightly differ in the specific events they generate.  Writing namespaces is particularly challenging, though there's certainly a way to solve those discrepancies, but again, it breaks a few ITs which are particularly sensitive to the exact XML generated.

Also, woodstox is 50% faster than the JDK implementation, so I definitely think we should use it.

File: maven-model/src/test/java/org/apache/maven/model/v4/StaxTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import org.apache.maven.api.model.InputSource;
+import org.apache.maven.api.model.Model;
+import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.Test;
+
+class StaxTest {
+
+    @Test
+    @Disabled
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-model/src/test/java/org/apache/maven/model/v4/StaxTest.java:39]: See below ...

File: maven-model-transform/pom.xml [hunk 2]
```diff
@@ -32,6 +32,14 @@ under the License.
       <groupId>org.codehaus.plexus</groupId>
       <artifactId>plexus-xml</artifactId>
     </dependency>
+    <dependency>
+      <groupId>org.codehaus.woodstox</groupId>
+      <artifactId>stax2-api</artifactId>
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-model-transform/pom.xml:37]: Yes, the round tripping (i.e. read the xml using stax and write it back to a string without any difference) is only supported by woodstox (the `stax2-api` is a mandatory dependency of woodstox anyway).

**Comment by gnodet** [comment]
> Depending on spaces in the prolog is a bug. Tests should be comparing XML to XML, not XML to strings. The latter is extremely brittle and can cause tests to break even in minor upgrades of a library. Alternately, we can canonicalize documents before comparing them. If you file issues on specific tests that do that, I can take a look.

That's not really the problem.  I could hack the tests.  However, the result xml is the one uploaded on central, and that one is ugly (because of the missing line breaks).  So I think this is important to keep.

File: maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java [hunk 3]
```diff
@@ -120,23 +120,29 @@ private Model read(Reader reader, Path pomFile, Map<String, ?> options) throws I
...
+        } catch (XMLStreamException e) {
+            Location location = e.getLocation();
+            throw new ModelParseException(
+                    e.getMessage(),
+                    location != null ? location.getLineNumber() : -1,
+                    location != null ? location.getColumnNumber() : -1,
+                    e);
         } catch (IOException e) {
             throw e;
         } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (gnodet) [reply maven-model-builder/src/main/java/org/apache/maven/model/io/DefaultModelReader.java:160]: The `IOException` can still occur during the call to `transformer.transform(...)` and we don't want to wrap it in another `IOException`...

**Comment by elharo** [comment]
I am sure there are ways to fix that.

**Comment by elharo** [comment]
Might or might not be related:


[INFO] 
Error:  Errors: 
Error:    DefaultMavenProjectBuilderTest.rereadPom_mng7063:334 � FileSystem C:\Users\RUNNER~1\AppData\Local\Temp\junit5723635972946138167\pom.xml: The process cannot access the file because it is being used by another process.

Error:    ProjectBuilderTest.testReadModifiedPoms(Path) � IO Failed to delete temp directory C:\Users\RUNNER~1\AppData\Local\Temp\junit3720659451607426598. The following paths could not be deleted (see suppressed exceptions for details): , child
[INFO] 
Error:  Tests run: 428, Failures: 0, Errors: 2, Skipped: 1
[INFO]

**Comment by gnodet** [comment]
I plan to provide additional support for xml namespaces, but I think a follow-up PR may be better.  I'll stop on that one.

**Comment by jira-importer** [comment]
Resolve #9088

**Pull request merged**